### DISCUSSION TITLE
Autenticação e Autorização ( Roles )

### DISCUSSION THREAD

**Discussion by BrunoCarolino7** [balta-io/forum#324] [Tira Dúvidas] [answered]
Olá, professor! Tenho uma dúvida... 

Curso ->  Fundamentos do ASP.NET 6
Módulo -> Autenticação e Autorização

Quando vamos passar papéis aos usuários, se ele é "user", "admin" ou "author", vi que isso é feito na mão inserindo na tabela de relacionamento (UserRole) o Id correspondente ao usuário e Id que corresponde a Role. 

Agora, minha dúvida é: como eu faria pra gerar esse papel automaticamente, digo, ao inserir um usuário X, esse usuario ser "user" e ao inserir um usuário Y esse usuário ser um "admin" sem que eu faça isso manualmente via Sql?

Preciso de uma luz, Balta, Por favor.  :)

Valeu !

**Comment by andrebaltieri** [comment]
Boa tarde, Bruno, como vai?

Você diz inserir na tabela UserRole os valores? Não sei se entendi direito a pergunta, mas vou tentar ajudar.

```csharp

var user = new User(...):

if(CONDIÇÃO PARA SER ADMIN)
{
// Recupera o perfil
var admin = await _context.Roles.FirstOrDefaultAsync(...);
user.Roles.Add(admin);
}else{
// Recupera o perfil
var userRole= await _context.Roles.FirstOrDefaultAsync(...);
user.Roles.Add(userRole);
}

await _context.SaveChangesAsync();
```

Seria isto?

**Reply by BrunoCarolino7** [reply]
Vamos indo :)

sim, inserir meio que automático. não sei se essa forma seria a mais correta, talvez poderia também puxar de uma tabela X e automaticamente puxando dessa tabela, o algoritmo já saber que é "admin" ou "user" etc.

![image](https://user-images.githubusercontent.com/91034533/236799480-53cbb623-b704-497d-8143-fc16dc7595d5.png)

Por exemplo, a ideia seria nesse endpoint criar um usuário só do tipo "user", já inserindo ele em UserRole como "user"

Não sei se fui claro :)

**Reply by andrebaltieri** [reply]
Bom dia, @BrunoCarolino7, como vai?

É uma saída sim, ou mesmo separar os endpoints, um para criar um usuário e outro um admin, imagino que o admin só possa ser criado por outros admins... ficaria mais fácil de gerenciar a autorização ao meu ver.

**Reply by BrunoCarolino7** [reply]
Acho que a pergunta certeira seria, como definir o papel (role) "user" já no endpoint por exemplo CreateUser ? Pergunto isso a respeito do token.. porque não entendi muito bem a questão do RoleClaimsExtension, porque pelo curso parece que você fez essa classe meio que ser generica. Poderia ao inves de deixar generico, gerar essas claims em cada Endpoint em si ? aí sim, no próprio endpoint CreateUser já inseri na claim o "user", daí pelo token já sabe q ele é do tipo "user".

Teria também que inserir essa informação no DB, certo ?

Teria como ser feito isso ? porque daí eu acho que resolve, talvez não seja a melhor opção, mas assim eu entendo.

**Reply by andrebaltieri** [reply]
Boa noite, @BrunoCarolino7, como vai? Não sei se entendi muito bem. Consegue detalhar o que está tentando fazer, um cenário mais real, assim posso ver qual seria a melhor forma de atender.

**Reply by BrunoCarolino7** [reply]
André,  Obrigado pela disposição em ajudar ! 

no curso você passa o papel pro usuário de forma manual no banco, relacionando o Id do usuário com Id da Role, assim, o usuário (Francisco) passa a ter um papel no sistema como (admin) por exemplo. Mas eu preciso que a própria api faça isso, que de alguma forma seja algo automatizado. Quando a pessoa for fazer o cadastro, ela automaticamente receba o papel de "user", como um usuário comum. Não queria que isso fosse passado na mão como é feito no curso. Deu pra entender rs ? 

Porque a controller atual de criar usuário no momento não cria com nenhum papel pra ele, no momento, eu crio um usuário sem papel e depois vou no banco e seto na unha a qual papel esse usuário desempenha, isso é inviável para aplicações reais e é essa minha dúvida, como poderia passar um papel pro usuário na hora de cria-lo. executar o endpoint ? 

(Isso seria top vc ensinar no curso que está por vir de autenticação e autorização), já rodei rodei no Yt etc e não encontrei nada... os conteúdos que encontrei os professores também passa os papéis dos usuários na unha pelo banco porque no geral parece que o conteúdo dos vídeos é ensinar a autorização do endpoint e não essa parte de passar o pepal pro usuário.

Enfim, aguardo sua ajuda, Balta ! Apesar de difícil de explicar, espero que tenha entendido rs 

Valeu até o momento !

![image](https://github.com/balta-io/forum/assets/91034533/ec69e5c8-a239-4354-9eb3-ddf0eb2c3ae3)

**Reply by andrebaltieri** [reply]
@BrunoCarolino7 neste caso, não seria só passar o ID do Role para API?
```
var role = context.Roles.FirstOrDefault(x => x.Id == model.RoleId);
if(role is not null)
   user.Roles.Add(role);
```

Veja se isto te dá um norte.

**Reply by BrunoCarolino7** [reply]
Opa, André ! Então, quando faço isso que você indicou, dá um erro:

![image](https://github.com/balta-io/forum/assets/91034533/ade09f9c-8330-4389-9f48-5acdf057a2a7)

na context ficou assim: 
![image](https://github.com/balta-io/forum/assets/91034533/da619d8a-a9f9-498c-b2cc-e087cca9d148)

aí o erro que dá ao tentar adicionar um tipo de role ao usuario como vc mostrou é este: 

![image](https://github.com/balta-io/forum/assets/91034533/3fd2d44d-5c88-4db6-a8a6-ac506178e5e2)

Se puder continuar a me ajudar neste saga kk

Valeu mesmo, André !

**Reply by andrebaltieri** [reply, marked as answer]
@BrunoCarolino7 este erro não seria por que sua coleção não está inicializada? Está tentando adicionar um item a uma lista nula?
